### PR TITLE
Explicitly support counterparty setting 0 channel reserve

### PR DESCRIPTION
**Pull request by TheBlueMatt**
A peer providing a channel_reserve_satoshis of 0 (or less than our
dust limit) is insecure, but only for them. Because some LSPs do it
with some level of trust of the clients (for a substantial UX
improvement), we explicitly allow it. Because its unlikely to
happen often in normal testing, we test it explicitly here.

Based on #1162, but mostly just so I don't have to fix silent merge conflicts in the tests.

### CODE REVIEW

**Review by valentinewallace — COMMENTED** [review]
I'm pretty much ACK! Cool to get this in

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -1703,12 +1717,8 @@ impl<Signer: Sign> Channel<Signer> {
 		if msg.channel_reserve_satoshis > self.channel_value_satoshis {
 			return Err(ChannelError::Close(format!("Bogus channel_reserve_satoshis ({}). Must not be greater than ({})", msg.channel_reserve_satoshis, self.channel_value_satoshis)));
 		}
-		if msg.channel_reserve_satoshis < self.holder_dust_limit_satoshis {
-			return Err(ChannelError::Close(format!("Peer never wants payout outputs? channel_reserve_satoshis was ({}). dust_limit is ({})", msg.channel_reserve_satoshis, self.holder_dust_limit_satoshis)));
-		}
-		let remote_reserve = Channel::<Signer>::get_holder_selected_channel_reserve_satoshis(self.channel_value_satoshis);
-		if msg.dust_limit_satoshis > remote_reserve {
-			return Err(ChannelError::Close(format!("Dust limit ({}) is bigger than our channel reserve ({})", msg.dust_limit_satoshis, remote_reserve)));
+		if msg.dust_limit_satoshis > self.holder_selected_channel_reserve_satoshis {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:1720]: Hmm, do we need to? I assume a `max_htlc_value_in_flight` of 21 million BTC is fine, we just (of course) won't send more than the channel value.

File: lightning/src/ln/channel.rs [hunk 2]
```diff
@@ -972,9 +987,6 @@ impl<Signer: Sign> Channel<Signer> {
 		if holder_selected_channel_reserve_satoshis < MIN_CHAN_DUST_LIMIT_SATOSHIS {
 			return Err(ChannelError::Close(format!("Suitable channel reserve not found. remote_channel_reserve was ({}). dust_limit_satoshis is ({}).", holder_selected_channel_reserve_satoshis, MIN_CHAN_DUST_LIMIT_SATOSHIS)));
 		}
-		if msg.channel_reserve_satoshis < MIN_CHAN_DUST_LIMIT_SATOSHIS {
-			return Err(ChannelError::Close(format!("channel_reserve_satoshis ({}) is smaller than our dust limit ({})", msg.channel_reserve_satoshis, MIN_CHAN_DUST_LIMIT_SATOSHIS)));
-		}
```
Comment (ariard) [inline lightning/src/ln/channel.rs:977]: If we detect a 0-sat reserve, should we at least log for the operator ? Can be interesting to verify that the channel reserve waiver is conduced with the right counterparty, even if in that case the security risk is on them. Maybe having a 0-reserve boolean in a future `ChannelStats` ?

**Review by ariard — COMMENTED** [review]
Code Review ACK c165359, comment isn't a blocker.

**Comment by TheBlueMatt** [comment]
Assigning @valentinewallace, post-re-ack I'll merge.

**Review by valentinewallace — APPROVED** [review]
ACK mod CI

**Comment by TheBlueMatt** [comment]
Oops lol, pushed a trivial fix.

**Review by ariard — COMMENTED** [review]
Code Review ACK c7bfed3b

**Comment by TheBlueMatt** [comment]
Squashed without changes, will land after CI:

```
```

**Comment by TheBlueMatt** [comment]
Fuzz CI failure is an upstream regression, not specific to this PR, will open a PR to resolve separately.

**Pull request merged**